### PR TITLE
fix(cli): Prevent stack overflow on deeply nested expressions

### PR DESCRIPTION
**Pull request by marschattha**
Suggested fix for https://github.com/qltysh/qlty/issues/1277

### CODE REVIEW

File: qlty-analysis/src/code/visitor.rs [hunk 1]
```diff
@@ -1,6 +1,9 @@
 use crate::Language;
+use tracing::warn;
 use tree_sitter::{Node, TreeCursor};
 
+const MAX_CURSOR_DEPTH: u32 = 400;
```
Comment (brynary) [inline qlty-analysis/src/code/visitor.rs:5]: How did you select 400?

**Review by brynary — COMMENTED** [review]
I agree that applying a maximum depth limit like this is necessary, otherwise I think there is no way to avoid stack overflows.

(I am a bit confused at why this does not occur when analyzing a single file, but I don't think we need to get to the bottom of that because this change looks good generally.)

File: qlty-analysis/src/code/visitor.rs (hunk 1 above)
Comment (marschattha) [reply qlty-analysis/src/code/visitor.rs:5]: Basically ran `qlty smells --all` a number of times on a folder containing these files and two others while logging the cursor depth, usually stack overflow occurred above 450 depth.

 We can lower it to be safer.

**Comment by marschattha** [comment]
> I am a bit confused at why this does not occur when analyzing a single file

My theory around that is that we are running in multiple threads and we are analyzing multiple files simultaneously for duplication checks?

File: qlty-analysis/src/code/visitor.rs (hunk 1 above)
Comment (brynary) [reply qlty-analysis/src/code/visitor.rs:5]: I think 400 is fine.

**Comment by brynary** [comment]
> My theory around that is that we are running in multiple threads and we are analyzing multiple files simultaneously for duplication checks?

Yes, we do use multiple threads, although I thought each gets its own stack allocation. Anyways, no

**Review by brynary — APPROVED** [review]
LGTM!

File: qlty-analysis/src/code/visitor.rs (hunk 1 above)
Comment (marschattha) [reply qlty-analysis/src/code/visitor.rs:5]: I think going to lower it a bit down to 300 to be safer on more systems.

**Pull request merged**